src/test/fuzz/block_index.cpp

```diff
@@ -0,0 +1,134 @@
+    files_info.reserve(files_count);
+    for (int i = 0; i < files_count; i++) {
+        if (auto file_info = ConsumeDeserializable<CBlockFileInfo>(fuzzed_data_provider)) {
+            files.push_back(std::make_unique<CBlockFileInfo>(std::move(*file_info)));
+            files_info.emplace_back(i, files.back().get());
+        } else {
+            return;
+        }
+    }
+
+    // Generate a number of block headers to be stored in the index.
+    int blocks_count = fuzzed_data_provider.ConsumeIntegralInRange(files_count * 10, files_count * 100);
+    std::vector<std::unique_ptr<CBlockIndex>> blocks;
+    blocks.reserve(blocks_count);
+    std::vector<const CBlockIndex*> blocks_info;
+    blocks_info.reserve(blocks_count);
+    for (int i = 0; i < blocks_count; i++) {
+        CBlockHeader header{ConsumeBlockHeader(fuzzed_data_provider)};
+        blocks.push_back(std::make_unique<CBlockIndex>(std::move(header)));
+        blocks.back()->phashBlock = &g_block_hash;
+        blocks_info.push_back(blocks.back().get());
+    }
+
+    // Store these files and blocks in the block index. It should not fail.
+    assert(block_index.WriteBatchSync(files_info, files_count - 1, blocks_info));
+
+    // We should be able to read every block file info we stored. Its value should correspond to
+    // what we stored above.
+    CBlockFileInfo info;
+    for (const auto& [n, file_info]: files_info) {
+        assert(block_index.ReadBlockFileInfo(n, info));
+        assert(info == *file_info);
+    }
+
+    // We should be able to read the last block file number. Its value should be consistent.
+    int last_block_file;
+    assert(block_index.ReadLastBlockFile(last_block_file));
+    assert(last_block_file == files_count - 1);
+
+    // We should be able to flip and read the reindexing flag.
+    bool reindexing;
+    block_index.WriteReindexing(true);
+    block_index.ReadReindexing(reindexing);
+    assert(reindexing);
+    block_index.WriteReindexing(false);
+    block_index.ReadReindexing(reindexing);
+    assert(!reindexing);
+
+    // We should be able to set and read the value of any random flag.
+    int flag_size = fuzzed_data_provider.ConsumeIntegralInRange(0, 100);
+    const std::string flag_name = fuzzed_data_provider.ConsumeBytesAsString(flag_size);
+    bool flag_value;
+    block_index.WriteFlag(flag_name, true);
+    block_index.ReadFlag(flag_name, flag_value);
+    assert(flag_value);
+    block_index.WriteFlag(flag_name, false);
+    block_index.ReadFlag(flag_name, flag_value);
+    assert(!flag_value);
+
+    // We should be able to load everything we've previously stored. Note to assert on the
+    // return value we need to make sure all blocks pass the pow check.
+    const auto params{Params().GetConsensus()};
+    const auto inserter = [&](const uint256&) {
+        return blocks.back().get();
+    };
+    WITH_LOCK(::cs_main, assert(block_index.LoadBlockIndexGuts(params, inserter, g_setup->m_interrupt)));
```

Comment:
Can we get rid of `g_setup` if we initialize a new `SignalInterrupt` here each interation?

Reply:
How about:
```diff
diff --git a/src/test/fuzz/block_index.cpp b/src/test/fuzz/block_index.cpp
index 57fc59e88b..ea3af83bd2 100644
--- a/src/test/fuzz/block_index.cpp
+++ b/src/test/fuzz/block_index.cpp
@@ -6 +6,2 @@
-#include <chainparams.h>
+#include <kernel/chainparams.h>
+#include <logging.h>
@@ -17 +18 @@ namespace {
-const BasicTestingSetup* g_setup;
+static const std::unique_ptr<const CChainParams> g_params{CChainParams::Main()};
@@ -20 +21 @@ const BasicTestingSetup* g_setup;
-uint256 g_block_hash;
+static const uint256 g_block_hash{g_params->GenesisBlock().GetHash()};
@@ -40 +41 @@ CBlockHeader ConsumeBlockHeader(FuzzedDataProvider& provider)
-    header.nBits = Params().GenesisBlock().nBits;
+    header.nBits = g_params->GenesisBlock().nBits;
@@ -49,3 +50,5 @@ void init_block_index()
-    static const auto testing_setup = MakeNoLogFileContext<>(ChainType::MAIN);
-    g_setup = testing_setup.get();
-    g_block_hash = Params().GenesisBlock().GetHash();
+    if (G_TEST_LOG_FUN) LogInstance().PushBackCallback(G_TEST_LOG_FUN);
+    LogInstance().m_print_to_file = false;
+    LogInstance().m_print_to_console = false;
+    LogInstance().m_log_timestamps = false;
+    LogInstance().StartLogging();
@@ -55,0 +59 @@ FUZZ_TARGET(block_index, .init = init_block_index)
+    auto interrupt = util::SignalInterrupt{};
@@ -128 +132 @@ FUZZ_TARGET(block_index, .init = init_block_index)
-    const auto params{Params().GetConsensus()};
+    const auto params{g_params->GetConsensus()};
@@ -132 +136,2 @@ FUZZ_TARGET(block_index, .init = init_block_index)
-    WITH_LOCK(::cs_main, assert(block_index.LoadBlockIndexGuts(params, inserter, g_setup->m_interrupt)));
+    WITH_LOCK(::cs_main, assert(block_index.LoadBlockIndexGuts(params, inserter, interrupt)));
+    assert(!interrupt);

```